src/coreclr/jit/helperexpansion.cpp

```diff
@@ -1553,3 +1553,263 @@ bool Compiler::fgVNBasedIntrinsicExpansionForCall_ReadUtf8(BasicBlock** pBlock,
     JITDUMP("ReadUtf8: succesfully expanded!\n")
     return true;
 }
+
+PhaseStatus Compiler::fgLateCastExpansion()
+{
+    if (!opts.IsOptimizedWithProfile())
+    {
+        return PhaseStatus::MODIFIED_NOTHING;
+    }
+
+    const bool preferSize = opts.jitFlags->IsSet(JitFlags::JIT_FLAG_SIZE_OPT);
+    if (preferSize)
+    {
+        // The optimization comes with a codegen size increase
+        JITDUMP("Optimized for size - bail out.\n");
+        return PhaseStatus::MODIFIED_NOTHING;
+    }
+    return fgExpandHelper<&Compiler::fgLateCastExpansionForCall>(true);
+}
+
+bool Compiler::fgLateCastExpansionForCall(BasicBlock** pBlock, Statement* stmt, GenTreeCall* call)
+{
+    if (!call->IsHelperCall() || !impIsCastHelperMayHaveProfileData(call->GetHelperNum()))
+    {
+        // Not a cast helper we're interested in
+        return false;
+    }
+
+    if ((call->gtCallMoreFlags & GTF_CALL_M_CAST_CAN_BE_EXPANDED) == 0)
+    {
+        // We either already expanded this cast helper or it's not eligible for expansion
+        return false;
+    }
```

Comment:
Is the flag really needed?

Reply:
> The current design of the helpercallexpansion will likely re-visit our call again so it will be an infinite loop otherwise.

It won't. I removed similar flag from the TLS expansion in #93830.

> Also, I am considering moving this phase after assertprop probably so then it will be part of OptRepeat

I think we should rather use `bbWeight`/`is rare` checks for this.

Reply:
> I think we should rather use bbWeight/is rare checks for this.

slow path is not guaranteed to be cold - it depends on profile (and for non-profiled casts it will be 50/50 except a few cases where we know that the slow path is only for throwing exceptions).

Reply:
(Alternatively, only do the helper expansion on the first iteration of the optimization loop, if we don't trust "is rare" enough)

Reply:
> (Alternatively, only do the helper expansion on the first iteration of the optimization loop, if we don't trust "is rare" enough)

yeah that will work, ok will remove the flag

Reply:
> slow path is not guaranteed to be cold - it depends on profile (and for non-profiled casts it will be 50/50 except a few cases where we know that the slow path is only for throwing exceptions).

Fair enough -- but I still think we should not try to do the helper expansion multiple times when we know there is nothing to expand, so I still think the flag can be removed.

I'm mainly interested in not having to go remove this flag again the next time I need a bit and we've run out.

Reply:
Ah, wait, I can't remove this flag yet - some casts are expanded in importer and some are in this new phase, I don't want a cast to be expanded in both phases. But I plan to move the cast expansion logic here completely so then I'll remove this flag. 

It's just that moving all of them here will likely require extra efforts in assertionprop etc since we're going to miss some branch opts in this late phase. It's not that important for profiled casts since there were enabled only recently so I will regress noone.

Reply:
Sounds good to me -- fine with keeping the flag if it's actually needed.